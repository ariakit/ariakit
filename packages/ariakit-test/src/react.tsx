import type { ReactElement } from "react";
import { StrictMode } from "react";
import * as ReactTestingLibrary from "@testing-library/react";
import { flushMicrotasks, nextFrame, wrapAsync } from "./__utils.js";

export * from "./index.js";

export interface RenderOptions
  extends Omit<ReactTestingLibrary.RenderOptions, "queries"> {
  strictMode?: boolean;
}

export async function render(ui: ReactElement, options?: RenderOptions) {
  const wrapper = (props: { children: ReactElement }) => {
    const Wrapper = options?.wrapper;
    const element = Wrapper ? <Wrapper {...props} /> : props.children;
    if (!options?.strictMode) return element;
    return <StrictMode>{element}</StrictMode>;
  };

  return wrapAsync(async () => {
    const { unmount } = ReactTestingLibrary.render(ui, { ...options, wrapper });
    await flushMicrotasks();
    await nextFrame();
    await flushMicrotasks();
    return unmount;
  });
}
