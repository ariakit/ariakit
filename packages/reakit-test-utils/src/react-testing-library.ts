export {
  act,
  fireEvent,
  render,
  wait,
  waitForDomChange,
  waitForElement,
  waitForElementToBeRemoved
} from "@testing-library/react";
