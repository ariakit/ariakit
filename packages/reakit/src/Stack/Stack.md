A [Box](../Box/Box.md) that stacks its children on top of each other. The first child element determines the size of the container, while the others positioned absolutely on top.

```jsx
import { Stack } from "reakit";

<Stack>
  <Box
    border="1px solid rgb(219, 112, 147)"
    width={80}
    height={80}
  />
  <Box
    backgroundColor="rgb(219, 112, 147)"
    border="1px solid rgb(219, 112, 147)"
    width={40}
    height={40}
  />
</Stack>
```

By default, the stacked children are anchored to the top left corner of the stack. This can be customized by setting the `anchor` prop to an array that may contain any combination of `"top"`, `"right"`, `"bottom"` and `"left"`. For example, `["bottom", "right"]` places all stacked children at the bottom right corner of the container. To stretch the children horizontally, include both `"left"` and `"right"` in the array; similarly, specifying both `"top"` and `"bottom"` causes the stacked children to be stretched vertically.

```jsx
import { Stack } from "reakit";

<Flex>
  <Stack anchor={['top', 'right']}>
    <Box
      border="1px solid rgb(219, 112, 147)"
      width={80}
      height={80}
    />
    <Box
      backgroundColor="rgb(219, 112, 147)"
      width={40}
      height={40}
    />
  </Stack>

  <Stack anchor={['bottom', 'right']} marginLeft={8}>
    <Box
      border="1px solid rgb(219, 112, 147)"
      width={80}
      height={80}
    />
    <Box
      backgroundColor="rgb(219, 112, 147)"
      width={40}
      height={40}
    />
  </Stack>

  <Stack anchor={['top', 'bottom', 'left']} marginLeft={8}>
    <Box
      border="1px solid rgb(219, 112, 147)"
      width={80}
      height={80}
    />
    <Box
      backgroundColor="rgb(219, 112, 147)"
      width={40}
    />
  </Stack>

  <Stack anchor={['top', 'left', 'right']} marginLeft={8}>
    <Box
      border="1px solid rgb(219, 112, 147)"
      width={80}
      height={80}
    />
    <Box
      backgroundColor="rgb(219, 112, 147)"
      height={40}
    />
  </Stack>
</Flex>
```
