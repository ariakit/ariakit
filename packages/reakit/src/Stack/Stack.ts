import { theme, withProp } from "styled-tools";
import styled from "../styled";
import use from "../use";
import Box, { BoxProps } from "../Box";

export type StackAnchor = "top" | "right" | "bottom" | "left";

export interface StackProps extends BoxProps {
  anchor?: StackAnchor[];
}

const ifAnchor = (anchor: StackAnchor, style: string) =>
  withProp("anchor", anchors => (anchors.includes(anchor) ? style : ""));

const Stack = styled(Box)<StackProps>`
  position: relative;

  > *:not(:first-child) {
    position: absolute;
    ${ifAnchor("top", "top: 0;")}
    ${ifAnchor("right", "right: 0;")}
    ${ifAnchor("bottom", "bottom: 0;")}
    ${ifAnchor("left", "left: 0;")}
  }

  ${theme("Stack")};
`;

Stack.defaultProps = {
  anchor: ["top", "left"]
};

export default use(Stack, "div");
