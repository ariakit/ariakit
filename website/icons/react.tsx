import type { SVGAttributes } from "react";

export function React(props: SVGAttributes<SVGSVGElement>) {
  return (
    <svg viewBox="-11.5 -10.23174 23 20.46348" {...props}>
      <circle
        className="fill-blue-600 dark:fill-[#61dafb]"
        cx="0"
        cy="0"
        r="2.05"
      />
      <g
        className="stroke-blue-600 dark:stroke-[#61dafb]"
        strokeWidth="1"
        fill="none"
      >
        <ellipse rx="11" ry="4.2" />
        <ellipse rx="11" ry="4.2" transform="rotate(60)" />
        <ellipse rx="11" ry="4.2" transform="rotate(120)" />
      </g>
    </svg>
  );
}
