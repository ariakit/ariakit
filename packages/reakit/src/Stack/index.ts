import Stack from "./Stack";

export * from "./Stack";

export default Stack;
