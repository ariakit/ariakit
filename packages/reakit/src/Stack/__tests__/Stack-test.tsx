import * as React from "react";
import { render } from "react-testing-library";
import Stack from "../Stack";

test("HTML attributes", () => {
  const { getByTestId } = render(
    <Stack id="test" aria-label="test" data-testid="test" />
  );

  expect(getByTestId("test")).toHaveAttribute("id", "test");
  expect(getByTestId("test")).toHaveAttribute("aria-label", "test");
});

test("Default", () => {
  const { container } = render(<Stack />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Anchor", () => {
  const { container } = render(<Stack anchor={["bottom", "right"]} />);
  expect(container.firstChild).toMatchSnapshot();
});
