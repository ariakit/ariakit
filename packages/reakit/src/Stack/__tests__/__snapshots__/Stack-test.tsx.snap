// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  position: relative;
}

.c0 > *:not(:first-child) {
  position: absolute;
  right: 0;
  bottom: 0;
}

<div
  class="c0 c1"
/>
`;

exports[`Default 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  position: relative;
}

.c0 > *:not(:first-child) {
  position: absolute;
  top: 0;
  left: 0;
}

<div
  class="c0 c1"
/>
`;
